Bluetooth: Audio: Factor out bt_audio_iso pool

MariuszSkamra opened the pull request:
This adds common bt_audio_iso pool that will be used across all the
roles/profiles. The pool range is dependent on the CONFIG_BT_ISO_MAX_CHAN
which is the maximum number of ISO connections the host can maintain.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: ascs: Move control point work to EP

Move the control point work to the EP. The aim is to cleanup the code so
that it there's no EP for ASE that means ASE is in idle state.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: Factor out bt_audio_iso pool

This adds common bt_audio_iso pool that will be used across all the
roles/profiles. The pool range is dependent on the CONFIG_BT_ISO_MAX_CHAN
which is the maximum number of ISO connections the host can maintain.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

fredrikdanebjer commented on subsys/bluetooth/audio/audio_iso.c:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	unicast = &unicast_pool[bt_audio_iso_index(audio_iso)];
+	__ASSERT_NO_MSG(unicast->acl == NULL);
```
I'm just wondering what the rational is for using `__ASSERT_NO_MSG` instead of `__ASSERT` ?

MariuszSkamra replied:
As far as I remember, assert prints the asserted line anyway. The check here is pretty straightforward, so I assumed additional message is not needed. Anyway, I changed the code so that there's no assert here anymore

Thalley reviewed (commented):
I did not look much into code changes, but just the overall change. 

The idea is to reduce the number of `bt_audio_iso` defined, right? 

If before we had `ASE_COUNT=2` and `CONFIG_BT_AUDIO_BROADCAST_SNK_COUNT=1` with `BROADCAST_SNK_STREAM_CNT=2`, then we would allocate 4 `bt_audio_iso` in total. 

If we only want to support 2 audio streams at any one time, then that is of course allocating 2 too many `bt_audio_iso`, so in that sense this change seems to make sense, as we would just set `CONFIG_BT_ISO_MAX_CHAN=2`, and thus save memory. 

Where this change starts to have challenges is if we want to support more ASEs than audio streams, since we at the QoS Configure state need to start coupling the ISO with the ASE. We may have to limit the number of ASEs to the number of ISO channels we support, but even then if we have 2 clients connected, then that won't work as ASEs is per client, and the number of `bt_audio_iso` isn't. The result of this will of course be that the stack will need to autonousmly reject requests to setting QoS, possibly without involving or notifying the upper layers. 

Another challenge is if we have a device that supports audio, but also uses ISO for other things (although I guess that's an unlikely situation). In that case, we will need to allocate more ISO channels (i.e. increase `CONFIG_BT_ISO_MAX_CHAN`), but this will just allocate additional `bt_audio_iso`s as well. Perhaps the number of `bt_audio_iso` should be controlled by a new Kconfig, or perhaps the MAX of broadcast source, broadcast sink, unicast client and unicast server support streams?

MariuszSkamra commented:
> I did not look much into code changes, but just the overall change.
> 
> The idea is to reduce the number of `bt_audio_iso` defined, right?
Yes, the idea is to reduce the memory usage.

> 
> If before we had `ASE_COUNT=2` and `CONFIG_BT_AUDIO_BROADCAST_SNK_COUNT=1` with `BROADCAST_SNK_STREAM_CNT=2`, then we would allocate 4 `bt_audio_iso` in total.
Assuming BT_CONN_MAX=1, yes.
In case BT_ISO_MAX_CHAN=2, BT_CONN_MAX=2 current code will allocate:
ASE_COUNT * BT_CONN_MAX + BT_AUDIO_BROADCAST_SNK_COUNT * BROADCAST_SNK_STREAM_CNT = 7.
So 5 additional `bt_audio_iso` that will not be used, because the ISO won't be created anyway.

for BT_CONN_MAX=3 it's 9 etc...

> Where this change starts to have challenges is if we want to support more ASEs than audio streams, since we at the QoS Configure state need to start coupling the ISO with the ASE. We may have to limit the number of ASEs to the number of ISO channels we support, but even then if we have 2 clients connected, then that won't work as ASEs is per client, and the number of `bt_audio_iso` isn't. The result of this will of course be that the stack will need to autonousmly reject requests to setting QoS, possibly without involving or notifying the upper layers.

I see no problem here. Host will not allow to create more ISO connections anyway, right? It's good that the request will be rejected earlier instead of pretending that everything is OK and the CIS Request will be accepted.

> Another challenge is if we have a device that supports audio, but also uses ISO for other things (although I guess that's an unlikely situation). In that case, we will need to allocate more ISO channels (i.e. increase `CONFIG_BT_ISO_MAX_CHAN`), but this will just allocate additional `bt_audio_iso`s as well. Perhaps the number of `bt_audio_iso` should be controlled by a new Kconfig, or perhaps the MAX of broadcast source, broadcast sink, unicast client and unicast server support streams?

I've updated the implementation, so that the `bt_audio_iso` pool size depends on the Kconfig options you mentioned.
Anyway, it requires some fixes in Kconfig as well, because It allows to set up all of the values to maximum.

Thalley commented:
> > Where this change starts to have challenges is if we want to support more ASEs than audio streams, since we at the QoS Configure state need to start coupling the ISO with the ASE. We may have to limit the number of ASEs to the number of ISO channels we support, but even then if we have 2 clients connected, then that won't work as ASEs is per client, and the number of `bt_audio_iso` isn't. The result of this will of course be that the stack will need to autonousmly reject requests to setting QoS, possibly without involving or notifying the upper layers.
> 
> I see no problem here. Host will not allow to create more ISO connections anyway, right? It's good that the request will be rejected earlier instead of pretending that everything is OK and the CIS Request will be accepted.

But do we need a way to include the upper layers somehow? The upper layers may want to terminate older CIS to use newer CIS. Of course since the application have determined `CONFIG_BT_ISO_MAX_CHAN` and know when an ASE gets Codec Configured, it may be able to handle that anyhow as it is currently. 

> 
> > Another challenge is if we have a device that supports audio, but also uses ISO for other things (although I guess that's an unlikely situation). In that case, we will need to allocate more ISO channels (i.e. increase `CONFIG_BT_ISO_MAX_CHAN`), but this will just allocate additional `bt_audio_iso`s as well. Perhaps the number of `bt_audio_iso` should be controlled by a new Kconfig, or perhaps the MAX of broadcast source, broadcast sink, unicast client and unicast server support streams?
> 
> I've updated the implementation, so that the `bt_audio_iso` pool size depends on the Kconfig options you mentioned. Anyway, it requires some fixes in Kconfig as well, because It allows to set up all of the values to maximum.

Sounds good. Since we can also (currently) register a single ISO server, ISO cannot be use (as a server anyhow) for more than LE Audio :) 
Someone from our application team also pointed out a related issue with the `listen` mechanism in `stream.c` that I've captured in https://github.com/zephyrproject-rtos/zephyr/issues/51552. 

This PR is related but slightly out of scope to fix those, but fixes in this PR may affect the issues listed in that GH issue, so please have a look.

MariuszSkamra commented on subsys/bluetooth/audio/audio_iso.h:
```diff
@@ -0,0 +1,62 @@
+	uint8_t source_path_cc[CONFIG_BT_CODEC_MAX_DATA_COUNT * CONFIG_BT_CODEC_MAX_DATA_LEN];
+
+	union {
+		struct {
+			struct bt_audio_stream *stream;
+		} broadcast;
+		struct {
+			struct bt_audio_stream *sink_stream;
+			struct bt_audio_stream *source_stream;
+		} unicast;
```
I found some issues having struct bt_audio_stream * instead of struct bt_audio_ep * inside of bt_audio_iso.
I'm about to change it, to store EP pointer here

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -366,60 +357,76 @@ static struct bt_audio_ep *unicast_client_ep_find(struct bt_conn *conn,
+		paired_stream = audio_iso->unicast.sink_stream;
+	}
+
+	if (stream != NULL || paired_stream == NULL) {
+		return false;
+	}
```
Not sure I follow why we are returning `false` if `paired_stream == NULL`. Can you elaborate?

Thalley reviewed (changes requested):
Some comments, otherwise looks very nice. I like that you've used the same concepts from `conn.c` to handle the ref counters and lookups.

MariuszSkamra commented on subsys/bluetooth/audio/audio_iso.h:
```diff
@@ -0,0 +1,62 @@
+	BT_AUDIO_ISO_TYPE_BROADCAST = BIT(1),
+	BT_AUDIO_ISO_TYPE = BT_AUDIO_ISO_TYPE_UNICAST |
+			    BT_AUDIO_ISO_TYPE_BROADCAST,
+} __packed;
```
I pack the enum to decrease it's size from 4 to 1 byte. I'll move sink_path_cc and source_path_cc to be placed after that, so this might save few bytes :)

Thalley replied:
Generally (at least in the BT subsystem) we only pack enums if they represent a value that we want to send over air (e.g. if we want an enum for opcodes that we need to ensure is 8-bit). 

In all other cases, we avoid packing it so that we can leave that up to the compiler

MariuszSkamra committed:
Bluetooth: audio: unicast_client: Remove unused functions

This removes unused functions from unicast_client.c

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: unicast_client: Inform all related EP's about ISO state

This fixes application callbacks not being called on ISO connection
state change.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented on subsys/bluetooth/audio/audio_iso.c:
```diff
@@ -0,0 +1,143 @@
+		if (!old) {
+			return NULL;
+		}
+	} while (!atomic_cas(&audio_iso->ref, old, old + 1));
```
> I wonder if it would make sense/possible to merge this part with `bt_conn_ref`.
> 
> Essentially what it is doing is
> 
> ```c
> 	atomic_val_t old;
> 
> 	/* Reference counter must be checked to avoid incrementing ref from
> 	 * zero, then we should return NULL instead.
> 	 * Loop on clear-and-set in case someone has modified the reference
> 	 * count since the read, and start over again when that happens.
> 	 */
> 	do {
> 		old = atomic_get(ref_counter);
> 
> 		if (!old) {
> 			return NULL;
> 		}
> 	} while (!atomic_cas(ref_counter, old, old + 1));
> ```
> 
> The type of the struct that hold the reference counter doesn't really matter, right?

What you mean by `type`?

This code does double-check of reference counter just in case there someone else called the ref/unref (modified the counter basically).

Thalley replied:
By `type` I simply meant that the reference counter handling doesn't care about whether it's a reference counter in a `bt_conn` or a `bt_audio_iso` struct, and that the common code to add a reference if, and only if, the counter is non-zero could be moved to a common reference counter function.

MariuszSkamra replied:
But isn't the lifetime of bt_conn different than bt_audio_iso? Having the bt_audio_iso in `listening` state, it has no conn as I understand correctly. Even if it's the same, it would require to provide some kind of `destroy` callback to bt_conn which would require major changes I guess,

MariuszSkamra commented on subsys/bluetooth/audio/audio_iso.c:
```diff
@@ -0,0 +1,143 @@
+
+#define ISO_POOL_SIZE \
+	MIN(CONFIG_BT_ISO_MAX_CHAN, \
+	    (IF_ENABLED(CONFIG_BT_ASCS, \
+			(CONFIG_BT_ASCS_ASE_SNK_COUNT + \
+			 CONFIG_BT_ASCS_ASE_SRC_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_UNICAST_CLIENT, \
+			(CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_COUNT * \
+			 CONFIG_BT_AUDIO_UNICAST_CLIENT_GROUP_STREAM_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SOURCE, \
+			(CONFIG_BT_AUDIO_BROADCAST_SRC_COUNT * \
+			 CONFIG_BT_AUDIO_BROADCAST_SRC_STREAM_COUNT)) + \
+	     IF_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SINK, \
+			(CONFIG_BT_AUDIO_BROADCAST_SNK_COUNT * \
+			 CONFIG_BT_AUDIO_BROADCAST_SNK_STREAM_COUNT)) + 0))
```
done

MariuszSkamra replied:
Yeah, I'll remove this patch from PR

Thalley replied:
Are we going to modify this?

MariuszSkamra commented:
I'm keeping this as draft still. Unidirectional CIS works OK, but in case of bidirectional I see some issues still. Will investigate it

sjanc commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -874,57 +830,54 @@ BT_CONN_CB_DEFINE(conn_cb) = {
-		}
+	iso = bt_audio_iso_find(audio_iso_find_func, &params);
+	if (iso) {
+		return iso;
```
hmm shouldn't this return extra reference?

MariuszSkamra replied:
You have right. I forgot to return extra reference from `bt_audio_iso_find`

sjanc commented on subsys/bluetooth/audio/broadcast_source.c:
```diff
@@ -229,14 +228,20 @@ static int bt_audio_broadcast_source_setup_stream(uint8_t index,
-	stream->iso->qos->rx = NULL;
-	bt_audio_codec_qos_to_iso_qos(stream->iso->qos->tx, qos);
-	bt_audio_codec_to_iso_path(stream->iso->qos->tx->path, codec);
+	bt_audio_codec_qos_to_iso_qos(iso->chan.qos->tx, qos);
```
iso is used after being unreffed

MariuszSkamra replied:
OK, I agree it looks odd, I'll fix it. The point is that `ep` still holds the reference. See my comment here https://github.com/zephyrproject-rtos/zephyr/pull/51513/files/8c16dd8d7044627cce12f1f02a010c97c8b93fd5#r1012635325

MariuszSkamra commented on subsys/bluetooth/audio/broadcast_sink.c:
```diff
@@ -930,18 +919,20 @@ static int bt_audio_broadcast_sink_setup_stream(uint8_t index,
+
+	bt_audio_iso_init(iso, &broadcast_sink_iso_ops);
+	bt_audio_iso_bind_ep(iso, ep);
+	bt_audio_iso_unref(iso);
```
`bt_audio_iso_new` returns reference (ref = 1).
`bt_audio_iso_bind_ep` takes extra reference (ref = 2).
The `bt_audio_iso_unref` here is to drop the reference to end up with only 1 reference held by `ep`

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -45,62 +46,27 @@
 	(_id > CONFIG_BT_ASCS_ASE_SNK_COUNT ? BT_UUID_ASCS_ASE_SRC : BT_UUID_ASCS_ASE_SNK)
 #define ASE_COUNT (CONFIG_BT_ASCS_ASE_SNK_COUNT + CONFIG_BT_ASCS_ASE_SRC_COUNT)
 
+BUILD_ASSERT(ASE_COUNT <= CONFIG_BT_ISO_MAX_CHAN,
+	     "ASCS total ASE count exceeds ISO BT_ISO_MAX_CHAN");
```
But not every ASE has to be in <= QoS state right? So the check will still be invalid then.

Thalley replied:
Not sure I follow your comment. 

I think it makes sense to restrict the total number of ASEs to the total number of `audio_iso`s.

MariuszSkamra replied:
Hmm, I'm starting to think that this assert is correct. If the server exposes e.g. 4 ASE it means that it can handle => 4 CIS connections. However, if the number of ACL connections is e.g. 3, the number of CIS and non-idle ASE would be still 4. It's because even if there are more clients connected the number of audio endpoints that can be in use will be the same.

Thalley replied:
Keep in mind that `ASE_COUNT` is the number of endpoints _per_ ACL connection. 

So if `ASE_COUNT=4` and `CONFIG_BT_MAX_CONN=3`, then we have a total of 12 ASEs

MariuszSkamra commented on subsys/bluetooth/audio/audio_iso.c:
```diff
@@ -0,0 +1,216 @@
+{
+	iso->chan.ops = ops;
+	iso->chan.qos = &iso->qos;
+	iso->chan.qos->rx = NULL;
+	iso->chan.qos->tx = NULL;
```
Oh, I can see, thanks!

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -486,13 +444,24 @@ static void ascs_iso_recv(struct bt_iso_chan *chan,
+	}
+
+	stream = ep->stream;
+	__ASSERT(stream, "No stream for ep %p", ep);
```
oh, I lost the context, I guess you meant `ascs_iso_sent` where the check is missing. Will fix that!

Thalley commented on subsys/bluetooth/audio/audio_iso.c:
```diff
@@ -0,0 +1,220 @@
+	if (ep->dir == BT_AUDIO_DIR_SINK) {
+		__ASSERT(iso->rx.ep == ep,
+			 "iso %p not bound with ep %p", iso, ep);
+		iso->rx.ep = NULL;
+		qos->rx = NULL;
+	} else  {
+		__ASSERT(iso->tx.ep == ep,
+			 "iso %p not bound with ep %p", iso, ep);
+		iso->tx.ep = NULL;
+		qos->tx = NULL;
```
```suggestion
		iso->rx.ep = NULL;
	} else  {
		__ASSERT(iso->tx.ep == ep,
			 "iso %p not bound with ep %p", iso, ep);
		iso->tx.ep = NULL;
```

MariuszSkamra replied:
Hmm, when I applied the changes broadcaster tests stopped working for me

Thalley replied:
Really? That's odd. The RX and TX pointer should be set by the `init` function, right? 

The pointers are basically constant, and are just pointing to structs within the `bt_audio_iso`, so the pointer shouldn't change at all.

MariuszSkamra replied:
ha! found it :)

MariuszSkamra committed:
Bluetooth: iso: Fix calling wrong function to terminate BIG

The bt_iso_chan contains vaild pointers for Tx and Rx QoS parameters.
This unifies the implementation between different roles.
This fixes checking ISO type that is done based on ISO type that
is kept inside of conn.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
